docs/src/interface-definition.md

```diff
@@ -0,0 +1,167 @@
+
+    # add your own definition
+    return zeros(AtomsCalculators.default_force_eltype, length(system))
+end
+```
+
+This creates both `forces` and `forces!`. `AtomsCalculators.default_force_eltype` is a type that can be used to allocate force data. You can also allocate for some other type.
+
+!!! note "Type definition under @generate_complement macro"
+    You need to use explicit definition of type when using
+    `@generate_complement` macro. `Main.MyType` is fine `MyType` is not!
+
+    You also need to define the type before macro call.
+
+Alternatively the definition could have been done with
+
+```julia
+struct MyOtherType
+end
+
+AtomsCalculators.@generate_complement function AtomsCalculators.forces!(f::AbstractVector, system, calculator::Main.MyOtherType; kwargs...)
+    @assert length(f) == length(system)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition
+    for i in eachindex(f)
+        f[i] = zero(AtomsCalculators.default_force_eltype)
+    end
+
+    return f
+end
+```
+
+## Other Automatically Generated Calls
+
+Many methods have optimized calls when energy and forces (and/or virial) are calculated. To allow access to these calls there are also calls
+
+- `energy_forces` for potential energy and allocating forces
+- `energy_forces!` for potential energy and non-allocating forces
+- `energy_forces_virial` for potential energy, allocating forces and virial
+- `energy_forces_virial!` for potential energy, non-allocating forces and virial
+
+These all are generated automatically, if you have defined the corresponding individual methods. The main idea here is that you can implement more efficient methods by your self.
+
+Example implementation
+
+```julia
+function AtomsCalculators.potential_energy_forces(system, calculator::MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition here
+    E = 0.0u"eV"
+    f = zeros(AtomsCalculators.default_force_eltype, length(system))
+    return (;
+        :energy => E,
+        :forces => f
+    )
+end
+```
+
+Defining this does not overload the corresponding non-allocating call - you need to do that separately.
+
+Output for the combination methods is defined to have keys `:energy`, `:forces` and `:virial`. You can access them with
+
+- `output[:energy]` for energy
+- `output[:forces]` for forces
+- `output[:virial]` for viral
+
+The type of the output can be [NamedTuple](https://docs.julialang.org/en/v1/base/base/#Core.NamedTuple), as was in the example above, [Dictionary](https://docs.julialang.org/en/v1/base/collections/#Dictionaries) or any structure that has the keys implemented. The reason for this is that this allows everyone to implement the performance functions without being restricted to certain output type, and to allow using `haskey` to check the output.
```

Comment:
I see - the point is that this way you allow both NamedTuple and Dict. I'm not sure about this. If it's a Dict then I can't splat it and write 
```
E, F, V = energy_forces_virial(...)
```
which I would really prefer.

Reply:
We could drop Dict support and define that it has to be NamedTuple. My thought here was that it could be used defined structure as well. But we could require that you can splat it and then it does not matter is it NamedTuple or some special structure.

Reply:
No need to change anything really, just add to the docs. If you think Dict might be important for some, then keep it and we discuss with the group. If you think it is really not needed then just add a line to the docs.